instrumentation/httpclient/src/main/java/io/opentelemetry/instrumentation/auto/httpclient/HttpClientInstrumentation.java

```diff
@@ -0,0 +1,182 @@
+            .and(takesArgument(1, named("java.net.http.HttpResponse$BodyHandler"))),
+        HttpClientInstrumentation.class.getName() + "$SendAsyncAdvice");
+
+    return transformers;
+  }
+
+  public static class SendAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void methodEnter(
+        @Advice.Argument(value = 0) HttpRequest httpRequest,
+        @Advice.Local("otelSpan") Span span,
+        @Advice.Local("otelScope") Scope scope,
+        @Advice.Local("otelCallDepth") Depth callDepth) {
+
+      callDepth = TRACER.getCallDepth();
+      if (callDepth.getAndIncrement() == 0) {
+        span = TRACER.startSpan(httpRequest);
+        if (span.getContext().isValid()) {
+          scope = TRACER.startScope(span, httpRequest);
+        }
+      }
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void methodExit(
+        @Advice.Return HttpResponse result,
+        @Advice.Thrown Throwable throwable,
+        @Advice.Local("otelSpan") Span span,
+        @Advice.Local("otelScope") Scope scope,
+        @Advice.Local("otelCallDepth") Depth callDepth) {
+
+      if (callDepth.decrementAndGet() == 0 && scope != null) {
+        scope.close();
+        if (throwable == null) {
+          TRACER.end(span, result);
+        } else {
+          TRACER.endExceptionally(span, result, throwable);
+        }
+      }
+    }
+  }
+
+  public static class SendAsyncAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void methodEnter(
+        @Advice.Argument(value = 0) HttpRequest httpRequest,
+        @Advice.Argument(value = 1, readOnly = false) BodyHandler<?> bodyHandler,
+        @Advice.Local("otelSpan") Span span,
+        @Advice.Local("otelScope") Scope scope,
+        @Advice.Local("otelCallDepth") Depth callDepth) {
+
+      callDepth = TRACER.getCallDepth();
+      if (callDepth.getAndIncrement() == 0) {
+        span = TRACER.startSpan(httpRequest);
+        bodyHandler = new TracingBodyHandler<>(bodyHandler, span);
+        if (span.getContext().isValid()) {
+          scope = TRACER.startScope(span, httpRequest);
+        }
+      }
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void methodExit(
+        @Advice.Return(readOnly = false, typing = Typing.DYNAMIC) CompletableFuture<?> future,
+        @Advice.Thrown Throwable throwable,
+        @Advice.Local("otelSpan") Span span,
+        @Advice.Local("otelScope") Scope scope,
+        @Advice.Local("otelCallDepth") Depth callDepth) {
+
+      if (callDepth.decrementAndGet() == 0 && scope != null) {
+        scope.close();
+        if (throwable != null) {
+          TRACER.endExceptionally(span, null, throwable);
+        } else {
+          // FIXME: next lines breaks interception
+          future =
+              future.whenComplete(
+                  (o, t) -> {
```

Comment:
@malafeev Have you tried running with debug logs? I'm guessing this instrumentation is "getting muzzled". Lambdas create additional anonymous classes, so I believe you need to add something like `packageName + HttpClientInstrumentation$1" to the helper classes list.

Because this is a bit magical I tend to avoid lambdas in instrumentation, and instead define a class that implements the interface (effectively same bytecode). Though I've found that even that doesn't always work...

https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1153/files#r481762196

Reply:
thanks @anuraaga, let me try

Reply:
it works !!!